Move Barrier module into the standard modules

daviditen opened the pull request:
Moved Barrier.chpl from test/parallel/taskPar/sungeun/barrier to the standard
modules.  Added chpldoc comments for the module and barrier records.  Added
Barrier.chpl to the MODULES_TO_DOCUMENT list in the modules/Makefile.

bradcray commented:
High level question:  What would cause one to use Barrier vs. sBarrier?  (and if we can come up with a good answer, can we put it at the top of the docs, maybe as part of providing a little "here's what you'll find below" paragraph at the top of the module?)

My first reaction was that there probably isn't a good reason, given that atomics tend to way outperform syncs.  And, as a user, I'd really like to just use a barrier without worrying about the implementation (i.e., the library should do whatever it needs to to give me a good one).

Then my second reaction was "Does the atomic barrier use busy-waiting?" (since atomics don't cause tasks to block/yield quite like sync variables do).  If so, that might suggest that the sync/single one should be the default.

If we can't come up with any rationale for how a user would choose between the two, then I'd be tempted to simplify this module down to just a single Barrier implementation.

bradcray commented:
Another high-level reaction: Assuming we keep both barrier types, it seems as though we may not want to bother documenting the methods on the abstract barrier class, given that the user can't really do much with them -- any docs would really be more for the developer, but the developer should be comfortable with reading the code.

OTOH, if the conclusion to the previous question is "let's just support one barrier", then, one might be tempted to do away with the abstract barrier type altogether.

(I was also tempted to suggest just nodoc-ing the abstract barrier record itself, but I suppose that could be confusing since the two implementations would refer to it as their parent record?)

bradcray commented on modules/standard/Barrier.chpl:
```diff
@@ -0,0 +1,199 @@
+
+       :rtype: bool
+     */
+    inline proc try() {
```
Stylistically, my memory was that we were preferring to declare return types in procedure headers (and perhaps not redundantly use the :rtype: field, similar to how you didn't redundantly mention 'n's type when documenting those arguments.  But I don't feel fully confident that we reached consensus on this, so maybe it's just my personal preference.

bradcray commented:
This seems like a case where it might be really nice to have an example of using a barrier in the documentation itself (we've talked about making start_test-able tests in the chpldoc documentation, but don't have that ability yet; even so, it seems like this interface is unlikely to change enough that the code will grow stale without testing.  I'm imagining something simple like the following:

    config const numTasks = here.maxTaskPar;
    var b = new Barrier(numTasks);
    
    coforall tid in 1..numTasks {
      writeln("Task ", tid, " is entering the barrier");
      b.barrier();
      writeln("Task ", tid, " is past the barrier");
    }

bradcray commented:
I could also imagine throwing a "future directions" section into the module header (if it could be offset in some nice way) noting that ultimately, we would like to add language-level support for creating teams of tasks and supporting collectives like barriers on them.  This can be viewed as a less-elegant means of achieving barriers between uncoordinated tasks in the meantime.

bradcray commented:
Do we have any idea how the performance of this barrier is?  Might we want to put in a status note saying that there are likely optimization opportunities in the implementation?  Or do we think that's not appropriate in a chpldoc setting?  (I'm unsure).

bradcray commented:
Other than those comments, this looks good to me!  (and none of these comments are show-stoppers, just things that I think would make it better).

Note that I reviewed by creating and reading the docs, browsing a few of Sung's tests that use barriers, and verifying that they work.  If you want me to do more than that, let me know.

Thanks!

ronawho commented:
>Then my second reaction was "Does the atomic barrier use busy-waiting?" (since atomics don't cause tasks to block/yield quite like sync variables do). If so, that might suggest that the sync/single one should be the default.

The atomic `waitFor()` does a `chpl_task_yield()`, so it won't be busy-waiting.

bradcray commented:
More busy than a sync, though, right?  I.e., if I had 99 tasks in a barrier and 1 still trying to get work done, wouldn't the sync implementation just park those 99 tasks until the last one was done while the atomics would spend a lot of time cycling between the tasks only to find they had nothing to do?

I guess for some workloads, it could be argued that this wouldn't matter.  But if that final task also yielded for some reason, it seems like it could get dragged down waiting for its turn to be scheduled again?

ronawho commented:
Yeah, probably more busy than a sync when the number of tasks is greater than the number of logical cores (that's the threshold for using condition variables to wait instead of loop with chpl_task_yield.)

At least that's true for fifo. I don't have a sense for qthreads since qthreads uses atomics to implement syncs.

bradcray commented:
Well, at the very least, this conversation makes me think that we should probably keep both barriers for now until we understand their performance characteristics better, though I do think that, ultimately, a user shouldn't have to worry about making this type of choice.

ronawho commented:
Think it's worth adding a multi-locale performance test?

daviditen commented:
@bradcray, I addressed your comments/questions except the first and last.  I think @ronawho addressed your first question in the same way I would have by saying the atomic version calls waitFor which has a loop with a chpl_task_yield, so it busy waits, but also yields the processor after every iteration.

Your last question was about performance.  I'm not sure what to compare the performance of these implementations to in order to say whether they perform well or not.  Compare vs. an MPI barrier? or OpenMP? or something else?

daviditen commented:
On 09/02/2015 03:36 PM, Elliot Ronaghan wrote:
>
> Think it's worth adding a multi-locale performance test?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/chapel-lang/chapel/pull/2416#issuecomment-137237238>.
>

Yes, but I'd like to wait until some time after feature freeze if that's 
OK with you.

ronawho commented:
Yup, I think waiting till feature freeze is fine.

I can help you with adding a multi-locale performance test to the xc overlay

bradcray commented:
I wasn't proposing doing a performance comparison per se, just curious whether our hunch was that it would perform well or poorly.  My hunch is that it will be "poor" at scale because most barriers will use log-tree communication patterns and I suspect this is a all-to-one communication pattern.  In asking about adding a performance note, I don't feel like we need to qualify/quantify the performance in any way -- I was just suggesting that it might be wise to put a "The current implementation is designed to be correct but not scalable/performance" type of disclaimer so that nobody is caught by surprise or complains about things that we already know about (and could head off with a one-liner in the docs).

daviditen commented:
OK, that makes sense.  I'll add a comment in that vein.

bradcray commented:
When we do write a barrier performance test, I think we'll want a single-locale test in addition to the multi-locale one Elliot proposed (with varying numbers of tasks, say numCores/2, numCores, numCores*2, numCores*4, numCores*10).  I think in the multi-locale world, the network will probably dominate (and the ability to use, or not, network atomics depending on the runtime layer...).

bradcray commented:
David, for some reason I wasn't seeing your intervening commits in this github view, but I just refreshed the page and see them now.  This is definitely above my bar now.  I wanted to note that I didn't test the example code I pasted above (and what you ended up with looked pretty similar), so hope that you did (because what are the odds that I actually got it right?  :)  ).

daviditen commented:
Thanks @bradcray and @ronawho.

bradcray commented:
@daviditen: The requirement to use 'with (ref b)' makes me wonder:  Is a plain-old-record the right/best datatype for the barrier?  Better might be:

* a class: because the barrier object has to be a unique thing and can't be copied by the compiler or things could go awry.  But, the major downside is that you'd have to delete it (ugh).
* a record with a single field which is a reference to said class.  This would take care of the delete problem (given appropriate constructor/destructor), and I believe wouldn't fall afoul of all the refcounting that arrays/domains do, nor the c-level issues that strings do.

Now that you've pointed this out, I worry a bit that a record is not a good solution because (a) you'd never want the compiler to put in copies of it and (b) you don't want it to be const-ish by default (whereas for the class-in-record solution, only the class reference would be copied and const, both of which are just fine).
